### PR TITLE
UCP/AM: Pack user hdr to the end of message

### PR DESCRIPTION
**Pull request by brminich**
## What
Pack user header to the end of UCP Active message.
Now AM will be constructed like `| AM HDR | PAYLOAD | USER HDR|`, instead of  `| AM HDR | USER HDR | PAYLOAD |`

## Why ?
1. It is needed for adding AM alignment option, which would guarantee specific alignment of AM payload on the receiver
2. To allow user calling `ucp_am_recv_data_nbx` for eager messages inside AM data callback. To support this, UCX has to initialize rdesc before passing the data to the AM data callback, which (with current alignment) would overwrite user header.

### CODE REVIEW

**Comment by brminich** [comment]
@dmitrygx, @rakhmets, @hoopoepg, can you please review?

**Review by rakhmets — APPROVED** [review]

**Comment by brminich** [comment]
@yosefe, can you please review?

**Review by dmitrygx — APPROVED** [review]

**Review by hoopoepg — APPROVED** [review]

File: src/ucp/core/ucp_am.c [hunk 1]
```diff
@@ -392,20 +392,24 @@ static UCS_F_ALWAYS_INLINE ssize_t
...
-                                            length - user_header_length);
+        /* Pack user header to the end of message/fragment */
+        user_hdr = UCS_PTR_BYTE_OFFSET(buffer, payload_length);
+        ucp_am_pack_user_header(user_hdr, req);
+    }
+
+    return user_header_length +
+           ucp_dt_pack(req->send.ep->worker, req->send.datatype,
+                       req->send.mem_type, buffer, req->send.buffer,
+                       &req->send.state.dt, payload_length);
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:412]: can u pls check if moving user_hdr affects performance (ucp_am_bw)

File: src/ucp/core/ucp_am.c [hunk 2]
```diff
@@ -1126,11 +1124,11 @@ ucp_am_invoke_cb(ucp_worker_h worker, ucp_am_hdr_t *am_hdr,
     if (ucs_likely(am_cb->flags & UCP_AM_CB_PRIV_FLAG_NBX)) {
         param.recv_attr = recv_flags;
         param.reply_ep  = reply_ep;
-        user_hdr        = user_hdr_length ?
-                          UCS_PTR_BYTE_OFFSET(am_hdr, am_hdr_length) : NULL;
+        user_hdr = user_hdr_length ? UCS_PTR_BYTE_OFFSET(data, data_length) :
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1127]: align = (ignore clang here)

File: src/ucp/core/ucp_am.c [hunk 3]
```diff
@@ -1313,26 +1308,29 @@ ucp_am_handle_unfinished(ucp_worker_h worker, ucp_recv_desc_t *first_rdesc,
...
                                      reply_ep, &reply_ep);
-    msg        = ucp_am_data_ptr(&first_hdr->super.super, sizeof(*first_hdr));
+    msg        = first_hdr + 1;
 
     status     = ucp_am_invoke_cb(worker, &first_hdr->super.super,
-                                  sizeof(*first_hdr), msg, first_hdr->total_size,
-                                  reply_ep, recv_flags | UCP_AM_RECV_ATTR_FLAG_DATA);
+                                  sizeof(*first_hdr), msg,
+                                  first_hdr->total_size +
+                                      first_hdr->super.super.header_length,
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1316]: pass header and payload sizes separately

File: src/ucp/core/ucp_am.c [hunk 4]
```diff
@@ -1382,17 +1383,24 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_am_long_first_handler,
      * Need to allocate a separate rdesc which would be in one contigious chunk
      * with data buffer. */
     first_rdesc = ucs_malloc(first_hdr->total_size + sizeof(ucp_recv_desc_t) +
-                             sizeof(*first_hdr),
+                                     sizeof(*first_hdr) + user_hdr_length,
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1386]: duplicate of 1362-1663?

File: src/ucp/core/ucp_am.c [hunk 5]
```diff
@@ -1410,9 +1418,10 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_am_long_first_handler,
 
     ucs_list_add_tail(&ep_ext->am.started_ams, &first_rdesc->am_first.list);
 
-    /* Note: copy first chunk of data together with header, which contains
+    /* Note: copy first chunk of data together with AM header, which contains
      * data needed to process other fragments. */
-    ucp_am_handle_unfinished(worker, first_rdesc, first_hdr, am_length, 0, ep);
+    ucp_am_handle_unfinished(worker, first_rdesc, first_hdr,
+                             am_length - user_hdr_length, 0, ep);
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1424]: already calculated in 1362?

File: src/ucp/core/ucp_am.c (hunk 4 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:1386]: no, on 1362 i calculate the data length in this message fragment.
Here I calculate the amount of memory needed for: the whole message, rdesc, user hdr, AM first hdr

File: src/ucp/core/ucp_am.c (hunk 5 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:1424]: Nope: here I calculate length of data in this fragment + AM first header

File: src/ucp/core/ucp_am.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/core/ucp_am.c:1384]: in both places, there is a calc of first_hdr->total_size + sizeof(*first_hdr) + user_hdr_length

File: src/ucp/core/ucp_am.c [hunk 6]
```diff
@@ -1178,7 +1177,8 @@ ucp_am_handler_common(ucp_worker_h worker, ucp_am_hdr_t *am_hdr, size_t hdr_size
     }
 
     status = ucp_am_invoke_cb(worker, am_hdr, hdr_size, data,
-                              total_length - hdr_size, reply_ep, recv_flags);
+                              total_length - (hdr_size + am_hdr->header_length),
```
Comment (yosefe) [inline src/ucp/core/ucp_am.c:1180]: bugfix?

File: src/ucp/core/ucp_am.c (hunk 1 above)
Comment (brminich) [reply src/ucp/core/ucp_am.c:412]: tested with #6330, no degradations observed

File: src/ucp/core/ucp_am.c (hunk 4 above)
Comment (yosefe) [reply src/ucp/core/ucp_am.c:1384]: according to the calculation, seems can do like this:
```
x         = first_hdr->total_size + sizeof(*first_hdr) + user_hdr_length
remaining = x - am_length

...
ucs_malloc(x + sizeof(ucp_recv_desc_t))
```

File: src/ucp/core/ucp_am.c (hunk 6 above)
Comment (yosefe) [reply src/ucp/core/ucp_am.c:1180]: ok

**Review by yosefe — APPROVED** [review]

**Pull request merged**